+Accessing types in polymorphic queries (#6227)
edgedb/edgedb: docs/edgeql/select.rst

    ...
  }

Accessing types in polymorphic queries
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> While the type of an object is displayed inside polymorphic queries, this
> is simply a convenience of the REPL and not a property that can be accessed.
> This is particularly noticeable if you cast an object to ``json``, making it
> impossible to determine the type if the query is polymorphic:
> 
> .. code-block:: edgeql-repl
> 
>     db> select Person limit 1;
>     {default::Villain {id: 6c22bdf0-5c03-11ee-99ff-dfaea4d947ce}}
>     db> select <json>Person limit 1;
>     {Json("{\"id\": \"6c22bdf0-5c03-11ee-99ff-dfaea4d947ce\"}")}

Reviewer: The way GitHub is diffing this makes it look a lot more different than it actually is. 😅 I'll explain what I was trying to accomplish here after the suggestion:
```suggestion
While the type of an object is displayed alongside the results of polymorphic
queries run in the REPL, this is simply a convenience of the REPL and not a
property that can be accessed. This is particularly noticeable if you cast an
object to ``json``, making it impossible to determine the type if the query is
polymorphic. First, the result of a query as the REPL presents it with type
annotations displayed:

.. code-block:: edgeql-repl

    db> select Person limit 1;
    {default::Villain {id: 6c22bdf0-5c03-11ee-99ff-dfaea4d947ce}}

Note the type ``default::Villain``, which is displayed for the user's
convenience but is not actually part of the data returned. This is the same
query with the result cast as ``json`` to show only the data returned:

.. code-block:: edgeql-repl

    db> select <json>Person limit 1;
    {Json("{\"id\": \"6c22bdf0-5c03-11ee-99ff-dfaea4d947ce\"}")}

.. note::

    We will continue to cast subesequent examples in this section to ``json``,
    not because this is required for any of the functionality being
    demonstrated, but to remove the convenience type annotations provided by
    the REPL and make it easier to see what data is actually being returned by
    the query.
```

- Clarify that the type annotation only happens in the REPL
- Split the examples with and without `json` cast to drive home the difference we're trying to call attention to
- Add a note that explains the continued use of `json` cast to any readers who might not understand and might wrongly assume you must cast to `json` to get type information back on a query.

The type of an object is found inside ``__type__`` which is a link that

> carries various information about a type, including its ``name``.

Reviewer: ```suggestion
carries various information about the object's type, including its ``name``.
```
I always fear there could be confusion between data type and object type, so I like to be explicit about which we're talking about.

.. code-block:: edgeql-repl

    db> select <json>Person { 
    ...  __type__: {
    ...    name 
    ...    } 
    ...  } limit 1;
    {Json("{\"__type__\": {\"name\": \"default::Villain\"}}")}

This information can be pulled into the top level by assigning a name to
the ``name`` property inside ``__type__``:

>     db> select <json>Person { n := .__type__.name } limit 1;
>     {Json("{\"n\": \"default::Villain\"}")}

Reviewer: ```suggestion
.. code-block:: edgeql-repl

    db> select <json>Person { type := .__type__.name } limit 1;
    {Json("{\"type\": \"default::Villain\"}")}
```

- Add `code-block` directive so it builds
- Change `n` to `type`. `type` feels more descriptive and since the `Person` type has a `name` property, `n` could be conflated to have something to do with that rather than the name of the type.

> There is nothing magical about ``__type__``; it is a simple link to an object
> called ``ObjectType`` that leads to all of the possible information to know
> about any type in the database. The splat operator can be used to see this
> object's makeup, while the double splat operator produces too much input to
> show on this page. Playing around with the splat and double splat operator
> inside ``__type__`` is a quick way to get some insight into the internals
> of EdgeDB.

Reviewer: Shifted this slightly to center more around this object's `__type__` link rather than about the `ObjectType` type since we're starting this description from `__type__`. Information is basically the same but the perspective has been shifted.
```suggestion
There is nothing magical about ``__type__``; it is a simple link to an object
of the type ``ObjectType`` which contains all of the possible information to
know about the type of the current object. The splat operator can be used to
see this object's makeup, while the double splat operator produces too much
output to show on this page. Playing around with the splat and double splat
operator inside ``__type__`` is a quick way to get some insight into the
internals of EdgeDB.
```
